src/OpenApi/Schema.hs

```diff
@@ -0,0 +1,208 @@
+module OpenApi.Schema where
+
+import Data.Aeson.Types (FromJSON, FromJSONKey, ToJSON, ToJSONKey)
+import Data.Map.Strict (Map)
+import Data.String (String)
+import Data.Tuple (snd, uncurry)
+import GHC.Generics
+import Numeric.Natural
+import OpenApi.JSON
+import OpenApi.Prelude
+import Prelude (undefined)
+
+import qualified Data.Aeson          as JSON
+import qualified Data.Aeson.Types    as JSON
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.List           as List
+import qualified Data.Map.Strict     as Map
+import qualified GHC.Enum            as GHC
+
+data AdditionalProperties
+  = AdditionalPropertiesBool Bool
+  | AdditionalPropertiesSchema Schema
+  deriving stock Show
+
+instance FromJSON AdditionalProperties where
+  parseJSON = \case
+    value@(JSON.Object _) -> AdditionalPropertiesSchema <$> JSON.parseJSON value
+    (JSON.Bool bool)      -> pure $ AdditionalPropertiesBool bool
+    value                 -> JSON.typeMismatch "Object or Bool" value
+
+instance ToJSON AdditionalProperties where
+  toJSON = \case
+    AdditionalPropertiesBool bool     -> JSON.toJSON bool
+    AdditionalPropertiesSchema schema -> JSON.toJSON schema
+
+newtype MaxLength = MaxLength Natural
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+newtype MaxProperties = MaxProperties Natural
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+newtype MinLength = MinLength Natural
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+newtype MinProperties = MinProperties Natural
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+data Properties = Properties (Map PropertyName Schema) | EmptyProperties
+  deriving stock Show
+
+instance FromJSON Properties where
+  parseJSON = JSON.withObject "properties" $ \map ->
+    if HashMap.null map
+      then pure EmptyProperties
+      else Properties <$> JSON.parseJSON (JSON.Object map)
+
+instance ToJSON Properties where
+  toJSON = \case
+    EmptyProperties -> JSON.object empty
+    Properties map  -> JSON.toJSON map
+
+newtype PropertyName = PropertyName Text
+  deriving newtype (Eq, FromJSON, FromJSONKey, Ord, ToJSON, ToJSONKey, ToText)
+  deriving stock   Show
+
+newtype MultipleOf = MultipleOf Natural
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+newtype ResourceId = ResourceId Text
+  deriving newtype (FromJSON, ToJSON)
+  deriving stock   Show
+
+newtype Name = Name Text
+  deriving newtype (Eq, FromJSONKey, Ord, ToJSON, ToJSONKey, ToText)
+  deriving stock   Show
+
+newtype Pattern = Pattern Text
+  deriving newtype (FromJSON, ToJSON, ToText)
+  deriving stock   Show
+
+data Schema = Content SchemaObject | Reference Name
+  deriving stock Show
+
+instance FromJSON Schema where
+  parseJSON = parseRefSum Name Content Reference "#/components/schemas/" "Schema"
+
+instance ToJSON Schema where
+  toJSON = \case
+    Content schemaObject
+      -> JSON.toJSON schemaObject
+    Reference name ->
```

Comment:
@mbj The `->` should be placed consistently in the two `case` legs here. I've no strong opinion about where it's better, though I'm sure you do, knowing where you'd typically place it.

Reply:
I foxed the instance above.